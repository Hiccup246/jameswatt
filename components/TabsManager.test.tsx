import TabsManager, {
  calcTabButtonTranslation,
  largestChildHeight,
} from "./TabsManager";
import { render, screen } from "@testing-library/react";

describe("TabsManager", () => {
  describe("When on a large device", () => {
    beforeAll(() => {
      Object.defineProperty(window, "matchMedia", {
        configurable: true,
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
          matches: false,
        })),
      });
    });

    it("renders correctly when given multiple children", () => {
      const { container } = render(
        <TabsManager>
          <div title="Panel 1" key="Panel 1">
            First Panel
          </div>
          <div title="Panel 2" key="Panel 2">
            Second Panel
          </div>
        </TabsManager>
      );

      expect(container).toMatchSnapshot();
    });

    it("renders the correct tab buttons", () => {
      Object.defineProperty(window, "matchMedia", {
        configurable: true,
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
          matches: false,
        })),
      });

      render(
        <TabsManager>
          <div title="Panel 1" key="Panel 1">
            First Panel
          </div>
          <div title="Panel 2" key="Panel 2">
            Second Panel
          </div>
        </TabsManager>
      );

      const tabButtonOne = screen.getByRole("button", { name: "Panel 1" });
      const tabButtonTwo = screen.getByRole("button", { name: "Panel 2" });

      expect(tabButtonOne).toBeInTheDocument();
      expect(tabButtonTwo).toBeInTheDocument();
    });

    it("renders the correct tab panels", () => {
      Object.defineProperty(window, "matchMedia", {
        configurable: true,
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
          matches: false,
        })),
      });

      render(
        <TabsManager>
          <div title="Panel 1" key="Panel 1">
            First Panel
          </div>
          <div title="Panel 2" key="Panel 2">
            Second Panel
          </div>
        </TabsManager>
      );

      const panelOneText = screen.getByText("First Panel");
      const panelTwoText = screen.getByText("Second Panel");

      expect(panelOneText).toBeInTheDocument();
      expect(panelTwoText).toBeInTheDocument();
    });
  });

  describe("When on a small device", () => {
    beforeAll(() => {
      Object.defineProperty(window, "matchMedia", {
        configurable: true,
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
          matches: true,
        })),
      });
    });

    it("renders correctly when given multiple children", () => {
      const { container } = render(
        <TabsManager>
          <div title="Panel 1" key="Panel 1">
            First Panel
          </div>
          <div title="Panel 2" key="Panel 2">
            Second Panel
          </div>
        </TabsManager>
      );

      expect(container).toMatchSnapshot();
    });

    it("renders the correct tab buttons", () => {
      Object.defineProperty(window, "matchMedia", {
        configurable: true,
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
          matches: false,
        })),
      });

      render(
        <TabsManager>
          <div title="Panel 1" key="Panel 1">
            First Panel
          </div>
          <div title="Panel 2" key="Panel 2">
            Second Panel
          </div>
        </TabsManager>
      );

      const tabButtonOne = screen.getByRole("button", { name: "Panel 1" });
      const tabButtonTwo = screen.getByRole("button", { name: "Panel 2" });

      expect(tabButtonOne).toBeInTheDocument();
      expect(tabButtonTwo).toBeInTheDocument();
    });

    it("renders the correct tab panels", () => {
      Object.defineProperty(window, "matchMedia", {
        configurable: true,
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
          matches: false,
        })),
      });

      render(
        <TabsManager>
          <div title="Panel 1" key="Panel 1">
            First Panel
          </div>
          <div title="Panel 2" key="Panel 2">
            Second Panel
          </div>
        </TabsManager>
      );

      const panelOneText = screen.getByText("First Panel");
      const panelTwoText = screen.getByText("Second Panel");

      expect(panelOneText).toBeInTheDocument();
      expect(panelTwoText).toBeInTheDocument();
    });
  });
});

describe("calcTabButtonTranslation", () => {
  describe("When on a large device", () => {
    beforeAll(() => {
      Object.defineProperty(window, "matchMedia", {
        configurable: true,
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
          matches: false,
        })),
      });
    });

    it("returns the correct Y translation when called with 5 and 8", () => {
      expect(calcTabButtonTranslation(5, 8)).toBe("translateY(8px)");
    });

    it("returns the correct Y translation when called with 0 and 0", () => {
      expect(calcTabButtonTranslation(0, 0)).toBe("translateY(0px)");
    });
  });

  describe("When on a small device", () => {
    beforeAll(() => {
      Object.defineProperty(window, "matchMedia", {
        configurable: true,
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
          matches: true,
        })),
      });
    });

    it("returns the correct X translation when called with 10 and 3", () => {
      expect(calcTabButtonTranslation(10, 3)).toBe("translateX(10px)");
    });

    it("returns the correct X translation when called with 0 and 0", () => {
      expect(calcTabButtonTranslation(0, 0)).toBe("translateX(0px)");
    });
  });
});

describe("largestChildHeight", () => {
  it("returns its default value when passed an empty element", () => {
    const mockElement: HTMLDivElement = document.createElement("div");
    expect(largestChildHeight(mockElement)).toBe(1000);
  });

  it("returns the height of the child element when given an element with one inner child", () => {
    const mockElement: HTMLDivElement = document.createElement("div");
    const mockChild: HTMLDivElement = document.createElement("div");
    jest.spyOn(mockChild, "clientHeight", "get").mockImplementation(() => 100);
    mockElement.appendChild(mockChild);

    expect(largestChildHeight(mockElement)).toBe(100);
  });

  it("returns the height of the largest child when given an element with multiple inner children", () => {
    const mockElement: HTMLDivElement = document.createElement("div");
    const mockChildOne: HTMLDivElement = document.createElement("div");
    const mockChildTwo: HTMLDivElement = document.createElement("div");

    jest
      .spyOn(mockChildOne, "clientHeight", "get")
      .mockImplementation(() => 100);
    mockElement.appendChild(mockChildOne);

    jest
      .spyOn(mockChildTwo, "clientHeight", "get")
      .mockImplementation(() => 200);
    mockElement.appendChild(mockChildTwo);

    expect(largestChildHeight(mockElement)).toBe(200);
  });
});
