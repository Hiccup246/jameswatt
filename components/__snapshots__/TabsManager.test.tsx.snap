// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabsManager When on a large device renders correctly when given multiple children 1`] = `
<div
  className="max-small:block flex w-full"
>
  <div
    className="
          max-small:flex max-small:w-full max-small:flex-row max-small:gap-0 max-small:overflow-y-hidden max-small:overflow-x-scroll max-small:py-0 max-small:pb-2 relative z-10 m-0 flex
          h-fit w-max list-none flex-col gap-6
          py-2.5
        "
  >
    <button
      aria-label="Panel 1"
      className="
                bg-transparent max-small:mb-0 max-small:shrink-0 max-small:grow-0 max-small:justify-center relative z-20 flex w-fit
                cursor-pointer whitespace-nowrap px-5 py-2
                focus-visible:border-none 
              "
      onClick={[Function]}
    >
      Panel 1
    </button>
    <button
      aria-label="Panel 2"
      className="
                bg-transparent max-small:mb-0 max-small:shrink-0 max-small:grow-0 max-small:justify-center relative z-20 flex w-fit
                cursor-pointer whitespace-nowrap px-5 py-2
                focus-visible:border-none 
              "
      onClick={[Function]}
    >
      Panel 2
    </button>
    <div
      className="
            transition-[transform width] ease-[cubic-bezier(0.645, 0.045, 0.355, 1)]
            max-small:bottom-2.5 max-small:top-auto visible absolute left-0 top-0 z-10 h-[40px] w-[122px] rounded
            bg-brown delay-100 duration-200 dark:bg-darkgrey
          "
      style={
        {
          "transform": "translateY(10px)",
        }
      }
    />
    <div
      className="
            max-small:hidden visible absolute left-0 top-0 ml-1
            h-full w-0.5
            rounded-sm bg-black dark:bg-softwhite
          "
    />
  </div>
  <div
    className="max-small:ml-0 relative ml-5 w-full"
  >
    <div
      className="
                absolute left-0 top-0 m-0 w-full p-0 transition-opacity duration-700 ease-in-out visible relative pt-2.5 opacity-100"
    >
      <div
        title="Panel 1"
      >
        First Panel
      </div>
    </div>
    <div
      className="
                absolute left-0 top-0 m-0 w-full p-0 transition-opacity duration-700 ease-in-out invisible opacity-0"
    >
      <div
        title="Panel 2"
      >
        Second Panel
      </div>
    </div>
  </div>
</div>
`;

exports[`TabsManager When on a small device renders correctly when given multiple children 1`] = `
<div
  className="max-small:block flex w-full"
>
  <div
    className="
          max-small:flex max-small:w-full max-small:flex-row max-small:gap-0 max-small:overflow-y-hidden max-small:overflow-x-scroll max-small:py-0 max-small:pb-2 relative z-10 m-0 flex
          h-fit w-max list-none flex-col gap-6
          py-2.5
        "
  >
    <button
      aria-label="Panel 1"
      className="
                bg-transparent max-small:mb-0 max-small:shrink-0 max-small:grow-0 max-small:justify-center relative z-20 flex w-fit
                cursor-pointer whitespace-nowrap px-5 py-2
                focus-visible:border-none 
              "
      onClick={[Function]}
    >
      Panel 1
    </button>
    <button
      aria-label="Panel 2"
      className="
                bg-transparent max-small:mb-0 max-small:shrink-0 max-small:grow-0 max-small:justify-center relative z-20 flex w-fit
                cursor-pointer whitespace-nowrap px-5 py-2
                focus-visible:border-none 
              "
      onClick={[Function]}
    >
      Panel 2
    </button>
    <div
      className="
            transition-[transform width] ease-[cubic-bezier(0.645, 0.045, 0.355, 1)]
            max-small:bottom-2.5 max-small:top-auto visible absolute left-0 top-0 z-10 h-[40px] w-[122px] rounded
            bg-brown delay-100 duration-200 dark:bg-darkgrey
          "
      style={
        {
          "height": "2px",
          "transform": "",
        }
      }
    />
    <div
      className="
            max-small:hidden visible absolute left-0 top-0 ml-1
            h-full w-0.5
            rounded-sm bg-black dark:bg-softwhite
          "
    />
  </div>
  <div
    className="max-small:ml-0 relative ml-5 w-full"
  >
    <div
      className="
                absolute left-0 top-0 m-0 w-full p-0 transition-opacity duration-700 ease-in-out visible relative pt-2.5 opacity-100"
    >
      <div
        title="Panel 1"
      >
        First Panel
      </div>
    </div>
    <div
      className="
                absolute left-0 top-0 m-0 w-full p-0 transition-opacity duration-700 ease-in-out invisible opacity-0"
    >
      <div
        title="Panel 2"
      >
        Second Panel
      </div>
    </div>
  </div>
</div>
`;
